### PR TITLE
Add support for new API Gateway service.

### PR DESCRIPTION
**Pull request by jthomas**
Supports passing token will switch to using provider implementation rather than built-in platform service.

Add unit tests and verified integrations still work against the current platform.

### CODE REVIEW

**Comment by jthomas** [comment]
Whoever merges this, can you also draft a new release?

v3.6.0 (12.05.2017)
- Add support for new API Gateway service.

File: README.md [hunk 1]
```diff
@@ -68,16 +68,19 @@ _Client constructor supports the following mandatory parameters:_
...
 - **ignore_certs**. Turns off server SSL/TLS certificate verification. This allows the client to be used against local deployments of OpenWhisk with a self-signed certificate. Defaults to false.
+- **apigw_token**. API Gateway service authentication token. This is mandatory for using an external API Gateway service, rather than the built-in api gateway.
+- **apigw_space_guid**. API Gateway space identifier. This is optional when using an API gateway service, defaults to the authentication uuid.
 
 ### environment variables
 
-Client constructor will read values for the `apihost`, `namespace` and `api_key` options from the environment if the following parameters are set. Explicit parameter values override these values.
+Client constructor will read values for the `apihost`, `namespace`, `api_key`, `apigw_token` and `apigw_space_guid` options from the environment if the following parameters are set. Explicit options have precedence over environment values.
 
 - *\_\_OW_API_HOST*
```
Comment (rabbah) [inline README.md]: why is it `\_` here

File: README.md [hunk 2]
```diff
@@ -414,9 +417,13 @@ The following optional parameters are supported:
 - `namespace` - set custom namespace for endpoint
 - `params` - JSON object containing parameters for the feed being invoked (default: `{}`)
 
-## api gateway (experimental)
+## api gateway
+
+OpenWhisk supports a [built-in API gateway service](https://github.com/openwhisk/openwhisk/blob/master/docs/apigateway.md) and external third-party providers.
```
Comment (rabbah) [inline README.md]: should probably update these links to apache/incubator-

File: lib/routes.js [hunk 3]
```diff
@@ -6,14 +6,27 @@ const names = require('./names')
 const CREATE_PARAMS = ['relpath', 'operation', 'action']
 
 class Routes extends BaseOperation {
-  static get routeMgmtApiPath () {
-    return 'experimental/web/whisk.system/routemgmt'
+  routeMgmtApiPath (path) {
+    return this.has_access_token() ?
+      `web/whisk.system/apimgmt/${path}.http` :
+      `experimental/web/whisk.system/routemgmt/${path}.json`
```
Comment (rabbah) [inline lib/routes.js]: should this just deprecate the support for `/experimental`?

File: test/integration/README.md [hunk 4]
```diff
@@ -3,13 +3,12 @@ Integrations Test
...
-    export __OW_NAMESPACE=<namespace to test against>
+    export __OW_API_KEY=<your api key>
+    export __OW_API_HOST=<openwhisk API hostname>
+    export __OW_NAMESPACE=<openwhisk namespace>
+    export __OW_APIGW_TOKEN=<api gateway token>
 
-You can retrieve these settings using the `wsk` CLI:
-
-    wsk property get --all
+You can retrieve these settings from the `.wskprops` file.
```
Comment (rabbah) [inline test/integration/README.md]: does the include the api gw token?

File: README.md (hunk 1 above)
Comment (jthomas) [inline README.md]: Good catch, the Markdown editor did this. I'll remove it.

File: README.md (hunk 2 above)
Comment (jthomas) [inline README.md]: Yep.

File: test/integration/README.md (hunk 4 above)
Comment (jthomas) [inline test/integration/README.md]: Yes, the `.wskprops` file stores the token.

File: lib/routes.js (hunk 3 above)
Comment (jthomas) [inline lib/routes.js]: I don't want to remove support for a feature that is still working within the platform.

**Pull request merged**